File: Data Strutures/Sorting/QuickSort.cpp
#include <bits/stdc++.h>
using namespace std;
#define FastIO ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);
#define int long long

int partition(int arr[], int low, int high)
{
	int pivot = arr[high];
	int i = low - 1;
	for(int j=low; j<high; j++)
	{
		if(arr[j] <= pivot)
		{
			i++;
			swap(arr[i], arr[j]);
		}
	}
	swap(arr[i+1], arr[high]);
	return i+1;
}

void quickSort(int arr[], int low, int high)
{
	if(low < high)
	{
		int pi = partition(arr, low, high);

		quickSort(arr, low, pi-1);
		quickSort(arr, pi+1, high);
	}
}

int print(int arr[], int n)
{
	for(int i=0; i<n; i++)
		cout<<arr[i]<<" ";
	cout<<"\n";
}

signed main()
{
    FastIO;
    int arr[10]={10, 9, 8, 7, 6, 5, 4, 3, 2, 1};

    print(arr, 10);

    quickSort(arr, 0, 9);
    
    print(arr, 10);
    
}

File: Data Strutures/Sorting/Java/QuickSort.java
public class QuickSort {
	static private int partition(int[] arr, int l, int h) {
		int i=l,j=l, p=h;
		for(i=l; i<h; i++) {
			if(arr[i] <= arr[p]) {
				int tmp = arr[i];
				arr[i] = arr[j];
				arr[j] = tmp;
				j+=1;
			}
		}
		
		int tmp = arr[h];
		arr[h] = arr[j];
		arr[j] = tmp;
				
		return j;		
	}
	static private void QuickSort(int[] arr, int l, int h) {
		if(l < h) {
			int p = partition(arr, l, h);
			QuickSort(arr, l, p-1);
			QuickSort(arr, p+1, h);
		}
	}
	
	static private void printArray(int[] arr) {
		for(int i=0; i<arr.length; i++)
			System.out.print(arr[i] + " ");
		System.out.println("\n");
	}
	
	public static void main(String[] args) {
		int[] arr = new int[10];
		for(int i=10; i>0; i--)
			arr[10-i] = i;
		printArray(arr);
		QuickSort(arr, 0, arr.length-1);
		printArray(arr);
	}
}